crates/rattler_solve/src/libsolv/mod.rs

```diff
@@ -0,0 +1,343 @@
+use std::collections::HashMap;
+use std::ffi::CString;
+
+mod ffi;
+mod keys;
+mod pool;
+mod queue;
+mod repo;
+mod solvable;
+mod solver;
+mod transaction;
+
+use pool::{Intern, Pool, Verbosity};
+use queue::Queue;
+
+use crate::libsolv::ffi::{
+    SOLVER_ERASE, SOLVER_FLAG_ALLOW_DOWNGRADE, SOLVER_FLAG_ALLOW_UNINSTALL, SOLVER_INSTALL,
+    SOLVER_SOLVABLE_PROVIDES, SOLVER_UPDATE,
+};
+use crate::{PackageOperation, RequestedAction, SolveError, SolverProblem};
+
+/// Convenience method that converts a string reference to a CString
+fn c_string<T: AsRef<str>>(str: T) -> CString {
+    CString::new(str.as_ref()).expect("should be convertable from string")
+}
+
+fn request_to_solvable_flags(action: RequestedAction) -> u32 {
```

Comment:
Could also turn this into a transparent type `SolvableFlags(u32)` and create a `From` implementation.